iceoryx_examples/icediscovery/iox_wait_for_service.cpp

```diff
@@ -0,0 +1,115 @@
+//! [include custom discovery]
+
+#include "iceoryx_hoofs/posix_wrapper/signal_watcher.hpp"
+#include "iceoryx_posh/runtime/posh_runtime.hpp"
+
+#include <iostream>
+
+using namespace discovery;
+
+constexpr char APP_NAME[] = "iox-wait-for-service";
+
+//! [service to wait for]
+iox::capro::IdString_t service{"Camera"};
+iox::capro::IdString_t instance{"FrontLeft"};
+iox::capro::IdString_t event{"Image"};
+//! [service to wait for]
+
+void printSearchResult(const discovery::ServiceContainer& result)
+{
+    std::cout << APP_NAME << " search result: " << (result.empty() ? "empty" : "") << std::endl;
+
+    for (auto entry : result)
+    {
+        std::cout << APP_NAME << " <" << entry.getServiceIDString() << ", " << entry.getInstanceIDString() << ", "
+                  << entry.getEventIDString() << ">" << std::endl;
+    }
+}
+
+Discovery* discoveryPtr{nullptr};
+
+void sigHandler(int)
+{
+    //! [unblock wait]
+    if (discoveryPtr)
+    {
+        discoveryPtr->unblockWait();
+    }
+    //! [unblock wait]
+}
+
+int main()
+{
+    iox::runtime::PoshRuntime::initRuntime(APP_NAME);
+
+    //! [create custom discovery]
+    // requires the runtime to be created first
+    Discovery discovery;
+    //! [create custom discovery]
+
+    discoveryPtr = &discovery;
+
+    auto sigTermGuard = iox::posix::registerSignalHandler(iox::posix::Signal::TERM, sigHandler);
+    auto sigIntGuard = iox::posix::registerSignalHandler(iox::posix::Signal::INT, sigHandler);
+
+    //! [define search query]
+    auto query = [&]() {
+        auto result = discovery.findService(service, instance, event);
+        return !result.empty();
+    };
+    //! [define search query]
+
+    std::cout << APP_NAME << " waiting for service <" << service << ", " << instance << ", " << event << "> ..."
+              << std::endl;
+
+    //! [wait until service was available]
+    bool serviceWasAvailable = discovery.waitUntil(query);
+    //! [wait until service was available]
+
+    // did we wake up due to an unblock or because the service was available?
+    if (serviceWasAvailable)
+    {
+        std::cout << APP_NAME << " <" << service << ", " << instance << ", " << event << "> was available" << std::endl;
+
+        // service was available, but we can never be sure the service is still available
+        // if this is important we need to monitor it (see discovery monitor example)
+
+        std::cout << APP_NAME << " waiting for any discovery change ..." << std::endl;
+
+        do
+        {
+            //! [wait until discovery changes]
+            discovery.waitUntilChange();
+            //! [wait until discovery changes]
+            std::cout << APP_NAME << " discovery changed. Searching <" << service << ", " << instance << ", " << event
+                      << "> ..." << std::endl;
+
+            //! [check service availability]
+            // loop while the service is available
+        } while (!discovery.findService(service, instance, event).empty());
```

Comment:
Maybe this should be made more clear by moving it in the loop body. More code but maybe better to understand.